Error when importing repo to BigTable

<!-- Please use this template while reporting a bug and provide as much info as possible.-->


**What happened**:

Observed the below error when importing a repo to BigTable:

```
com.google.api.gax.rpc.UnavailableException: io.grpc.StatusRuntimeException: UNAVAILABLE: A DropRowRange operation is already ongoing.
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:112)
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:41)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:86)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:84)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1127)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1286)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1055)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:807)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:578)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:548)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:541)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:567)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:71)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:735)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:716)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient.dropRowRange(BigtableTableAdminClient.java:736)
		at org.projectnessie.versioned.storage.bigtable.BigTableBackend.eraseRepositoriesAdminClient(BigTableBackend.java:169)
		at org.projectnessie.versioned.storage.bigtable.BigTableBackend.eraseRepositories(BigTableBackend.java:157)
		at org.projectnessie.versioned.storage.bigtable.BigTablePersist.erase(BigTablePersist.java:547)
		at org.projectnessie.versioned.storage.cache.CachingPersistImpl.erase(CachingPersistImpl.java:207)
		at org.projectnessie.quarkus.cli.ImportRepository.callWithPersist(ImportRepository.java:189)
		at org.projectnessie.quarkus.cli.BaseCommand.call(BaseCommand.java:55)
		at org.projectnessie.quarkus.cli.BaseCommand.call(BaseCommand.java:33)
		at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
		at picocli.CommandLine.access$1500(CommandLine.java:148)
		at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
		at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
		at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
		at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
		at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
		at io.quarkus.picocli.runtime.PicocliRunner$EventExecutionStrategy.execute(PicocliRunner.java:26)
		at picocli.CommandLine.execute(CommandLine.java:2170)
		at io.quarkus.picocli.runtime.PicocliRunner.run(PicocliRunner.java:40)
		at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:132)
		at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
		at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
		at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: A DropRowRange operation is already ongoing.
	at io.grpc.Status.asRuntimeException(Status.java:539)
	... 14 more
```

**What you expected to happen**:

No errors.

**How to reproduce it (as minimally and precisely as possible)**:

I will work on getting a clean reproducer but the error doesn't manifest deterministically it seems.

**Details**:
- Nessie server type (docker/built from source) and version: 0.65
- Client type (Ex: UI/Spark/pynessie ...) and version:

--- a/versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTableBackend.java
+++ b/versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTableBackend.java
@@ -26,6 +26,8 @@ import static org.projectnessie.versioned.storage.bigtable.BigTableConstants.TAB
 import static org.projectnessie.versioned.storage.bigtable.BigTableConstants.TABLE_REFS;
 
 import com.google.api.gax.rpc.NotFoundException;
+import com.google.api.gax.rpc.ResourceExhaustedException;
+import com.google.api.gax.rpc.UnavailableException;
 import com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient;
 import com.google.cloud.bigtable.admin.v2.models.CreateTableRequest;
 import com.google.cloud.bigtable.data.v2.BigtableDataClient;
@@ -166,8 +168,16 @@ final class BigTableBackend implements Backend {
 
     for (String repoId : repositoryIds) {
       ByteString prefix = copyFromUtf8(repoId + ':');
-      tableAdminClient.dropRowRange(tableRefs, prefix);
-      tableAdminClient.dropRowRange(tableObjs, prefix);
+      try {
+        tableAdminClient.dropRowRange(tableRefs, prefix);
+        tableAdminClient.dropRowRange(tableObjs, prefix);
+      } catch (ResourceExhaustedException e) {
+        LOGGER.warn("DropRowRange quota exceeded, trying the non-admin path", e);
+        return false;
+      } catch (UnavailableException e) {
+        LOGGER.warn("DropRowRange operation unavailable, trying the non-admin path", e);
+        return false;
+      }
     }
 
     return true;

--- a/versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTableBackend.java
+++ b/versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTableBackend.java
@@ -37,6 +37,8 @@ import com.google.cloud.bigtable.data.v2.models.Mutation;
 import com.google.cloud.bigtable.data.v2.models.Query;
 import com.google.cloud.bigtable.data.v2.models.Row;
 import com.google.protobuf.ByteString;
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
@@ -156,40 +158,40 @@ final class BigTableBackend implements Backend {
 
   @Override
   public void eraseRepositories(Set<String> repositoryIds) {
-    if (!eraseRepositoriesAdminClient(repositoryIds)) {
-      eraseRepositoriesNoAdminClient(repositoryIds);
-    }
+    List<String> toDelete = new ArrayList<>(repositoryIds);
+    eraseRepositoriesAdminClient(toDelete);
+    eraseRepositoriesNoAdminClient(toDelete);
   }
 
-  private boolean eraseRepositoriesAdminClient(Set<String> repositoryIds) {
-    if (tableAdminClient == null) {
-      return false;
-    }
-
-    for (String repoId : repositoryIds) {
-      ByteString prefix = copyFromUtf8(repoId + ':');
-      try {
-        tableAdminClient.dropRowRange(tableRefs, prefix);
-        tableAdminClient.dropRowRange(tableObjs, prefix);
-      } catch (ResourceExhaustedException e) {
-        LOGGER.warn("DropRowRange quota exceeded, trying the non-admin path", e);
-        return false;
-      } catch (UnavailableException e) {
-        LOGGER.warn("DropRowRange operation unavailable, trying the non-admin path", e);
-        return false;
+  private void eraseRepositoriesAdminClient(List<String> repositoryIds) {
+    if (tableAdminClient != null) {
+      while (!repositoryIds.isEmpty()) {
+        String repoId = repositoryIds.get(0);
+        ByteString prefix = copyFromUtf8(repoId + ':');
+        try {
+          tableAdminClient.dropRowRange(tableRefs, prefix);
+          tableAdminClient.dropRowRange(tableObjs, prefix);
+        } catch (ResourceExhaustedException e) {
+          LOGGER.warn("DropRowRange quota exceeded, trying the non-admin path", e);
+          return;
+        } catch (UnavailableException e) {
+          LOGGER.warn("DropRowRange operation unavailable, trying the non-admin path", e);
+          return;
+        }
+        repositoryIds.remove(repoId);
       }
     }
-
-    return true;
   }
 
-  private void eraseRepositoriesNoAdminClient(Set<String> repositoryIds) {
-    List<ByteString> prefixes =
-        repositoryIds.stream()
-            .map(repoId -> copyFromUtf8(repoId + ':'))
-            .collect(Collectors.toList());
-    eraseRepositoriesTable(tableRefs, prefixes);
-    eraseRepositoriesTable(tableObjs, prefixes);
+  private void eraseRepositoriesNoAdminClient(Collection<String> repositoryIds) {
+    if (!repositoryIds.isEmpty()) {
+      List<ByteString> prefixes =
+          repositoryIds.stream()
+              .map(repoId -> copyFromUtf8(repoId + ':'))
+              .collect(Collectors.toList());
+      eraseRepositoriesTable(tableRefs, prefixes);
+      eraseRepositoriesTable(tableObjs, prefixes);
+    }
   }
 
   private void eraseRepositoriesTable(String tableId, List<ByteString> prefixes) {

--- a/versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTableBackend.java
+++ b/versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTableBackend.java
@@ -25,9 +25,8 @@ import static org.projectnessie.versioned.storage.bigtable.BigTableConstants.MAX
 import static org.projectnessie.versioned.storage.bigtable.BigTableConstants.TABLE_OBJS;
 import static org.projectnessie.versioned.storage.bigtable.BigTableConstants.TABLE_REFS;
 
+import com.google.api.gax.rpc.ApiException;
 import com.google.api.gax.rpc.NotFoundException;
-import com.google.api.gax.rpc.ResourceExhaustedException;
-import com.google.api.gax.rpc.UnavailableException;
 import com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient;
 import com.google.cloud.bigtable.admin.v2.models.CreateTableRequest;
 import com.google.cloud.bigtable.data.v2.BigtableDataClient;
@@ -171,11 +170,8 @@ final class BigTableBackend implements Backend {
         try {
           tableAdminClient.dropRowRange(tableRefs, prefix);
           tableAdminClient.dropRowRange(tableObjs, prefix);
-        } catch (ResourceExhaustedException e) {
-          LOGGER.warn("DropRowRange quota exceeded, trying the non-admin path", e);
-          return;
-        } catch (UnavailableException e) {
-          LOGGER.warn("DropRowRange operation unavailable, trying the non-admin path", e);
+        } catch (ApiException e) {
+          LOGGER.warn("Could not erase repo with admin client, switching to data client", e);
           return;
         }
         repositoryIds.remove(repoId);

--- a/versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTableBackend.java
+++ b/versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTableBackend.java
@@ -36,10 +36,11 @@ import com.google.cloud.bigtable.data.v2.models.Mutation;
 import com.google.cloud.bigtable.data.v2.models.Query;
 import com.google.cloud.bigtable.data.v2.models.Row;
 import com.google.protobuf.ByteString;
-import java.util.ArrayList;
+import java.util.ArrayDeque;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Queue;
 import java.util.Set;
 import java.util.stream.Collectors;
 import javax.annotation.Nonnull;
@@ -157,15 +158,15 @@ final class BigTableBackend implements Backend {
 
   @Override
   public void eraseRepositories(Set<String> repositoryIds) {
-    List<String> toDelete = new ArrayList<>(repositoryIds);
+    Queue<String> toDelete = new ArrayDeque<>(repositoryIds);
     eraseRepositoriesAdminClient(toDelete);
     eraseRepositoriesNoAdminClient(toDelete);
   }
 
-  private void eraseRepositoriesAdminClient(List<String> repositoryIds) {
+  private void eraseRepositoriesAdminClient(Queue<String> repositoryIds) {
     if (tableAdminClient != null) {
       while (!repositoryIds.isEmpty()) {
-        String repoId = repositoryIds.get(0);
+        String repoId = repositoryIds.peek();
         ByteString prefix = copyFromUtf8(repoId + ':');
         try {
           tableAdminClient.dropRowRange(tableRefs, prefix);
@@ -174,7 +175,7 @@ final class BigTableBackend implements Backend {
           LOGGER.warn("Could not erase repo with admin client, switching to data client", e);
           return;
         }
-        repositoryIds.remove(repoId);
+        repositoryIds.poll();
       }
     }
   }

